packages/vite/src/node/optimizer/scan.ts

```diff
@@ -241,11 +263,31 @@ function esbuildScanPlugin(
             if (lang === 'ts' || lang === 'tsx' || lang === 'jsx') {
               loader = lang
             }
+            const srcMatch = openTag.match(srcRE)
             if (srcMatch) {
               const src = srcMatch[1] || srcMatch[2] || srcMatch[3]
               js += `import ${JSON.stringify(src)}\n`
             } else if (content.trim()) {
-              js += content + '\n'
+              // There can be module scripts (`<script context="module">` in Svelte and `<script>` in Vue)
+              // or local scripts (`<script>` in Svelte and `<script setup>` in Vue)
+              // We need to handle these separately in case variable names are reused between them
+              const contextMatch = openTag.match(contextRE)
+              const context =
+                contextMatch &&
+                (contextMatch[1] || contextMatch[2] || contextMatch[3])
+              if (
+                (path.endsWith('.vue') && setupRE.test(raw)) ||
+                (path.endsWith('.svelte') && context !== 'module')
+              ) {
+                const id = `virtual-module:${path}`
+                moduleScripts[id] = {
+                  loader,
+                  contents: content
+                }
+                js += `import '${id}';\n`
+              } else {
+                js += content + '\n'
+              }
```

Comment:
Isn't this condition inverted from what Evan proposed for the Vue case? 
Apply the new virtual module logic to (2) "Local" context script tags: <script setup> in vue, and <script> in svelte

Reply:
Doesn't this do it in the manner Evan suggested? The `setupRE` checks if it's a Vue `<script setup>` and the `context !== 'module'` checks if it's a `<script>` in Svelte. If either of those conditions are true then it creates the virtual module

Reply:
Sorry for the noise, I missread setupRE there 👍🏼